src/asynchronous/client.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2020 Ant Financial
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use nix::unistd::close;
+use protobuf::{CodedInputStream, CodedOutputStream, Message};
+use std::collections::HashMap;
+use std::os::unix::io::RawFd;
+use std::sync::{Arc, Mutex};
+
+use crate::common::MESSAGE_TYPE_RESPONSE;
+use crate::error::{Error, Result};
+use crate::ttrpc::{Code, Request, Response};
+
+use crate::asynchronous::stream::{receive, to_req_buf};
+use crate::r#async::utils;
+use tokio::{
+    self,
+    io::split,
+    prelude::*,
+    sync::mpsc::{channel, Receiver, Sender},
+    sync::Notify,
+};
+
+type RequestSender = Sender<(Vec<u8>, Sender<Result<Vec<u8>>>)>;
+type RequestReceiver = Receiver<(Vec<u8>, Sender<Result<Vec<u8>>>)>;
+
+type ResponseSender = Sender<Result<Vec<u8>>>;
+type ResponseReceiver = Receiver<Result<Vec<u8>>>;
+
+#[derive(Clone)]
+pub struct Client {
+    req_tx: RequestSender,
+}
+
+impl Client {
+    /// Initialize a new [`Client`].
+    pub fn new(fd: RawFd) -> Client {
+        let stream = utils::new_unix_stream_from_raw_fd(fd);
+
+        let (mut reader, mut writer) = split(stream);
+        let (req_tx, mut rx): (RequestSender, RequestReceiver) = channel(100);
+
+        let req_map = Arc::new(Mutex::new(HashMap::new()));
+        let req_map2 = req_map.clone();
+
+        let notify = Arc::new(Notify::new());
+        let notify2 = notify.clone();
+
+        // Request sender
+        tokio::spawn(async move {
+            let mut stream_id: u32 = 1;
+
+            while let Some((body, mut resp_tx)) = rx.recv().await {
+                let current_stream_id = stream_id;
+                stream_id += 2;
+
+                {
+                    let mut map = req_map2.lock().unwrap();
+                    map.insert(current_stream_id, resp_tx.clone());
+                }
+
+                let buf = to_req_buf(current_stream_id, body);
+                if let Err(e) = writer.write_all(&buf).await {
```

Comment:
if write failed， which means the connection was broken, should it break the loop and notify receiver to terminate.

Reply:
done